v2/cmd/gen/gen.go

```diff
@@ -0,0 +1,234 @@
+	}
+}
+
+func emitCommand() *cobra.Command {
+	var (
+		typ   string
+		out   string
+		start int
+		limit int
+	)
+	cmd := &cobra.Command{
+		Use:   "emit",
+		Short: "emit generated changesets to disk",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			itr, err := getChangesetIterator(typ)
+			if err != nil {
+				return err
+			}
+			ctx := core.Context{Context: cmd.Context()}
+
+			stream := compact.StreamingContext{
+				In:          make(chan compact.Sequenced),
+				Context:     ctx,
+				OutDir:      out,
+				MaxFileSize: 100 * 1024 * 1024,
+			}
+
+			var wg sync.WaitGroup
+			wg.Add(1)
+			go func() {
+				stats, err := stream.Compact()
+				if err != nil {
+					log.Fatal().Err(err).Msg("failed to compact")
+				}
+				log.Info().Msgf(stats.Report())
+				wg.Done()
+			}()
+
+			var cnt int64
+			for ; itr.Valid(); err = itr.Next() {
+				if err != nil {
+					return err
+				}
+				if limit > 0 && itr.Version() > int64(limit) {
+					break
+				}
+				nodes := itr.Nodes()
+				for ; nodes.Valid(); err = nodes.Next() {
+					cnt++
+
+					if itr.Version() < int64(start) {
+						if cnt%5_000_000 == 0 {
+							log.Info().Msgf("fast forward version=%d nodes=%s", itr.Version(), humanize.Comma(cnt))
+						}
+						continue
+					}
+
+					if cnt%500_000 == 0 {
+						log.Info().Msgf("version=%d nodes=%s", itr.Version(), humanize.Comma(cnt))
+					}
+
+					select {
+					case <-cmd.Context().Done():
+						close(stream.In)
+						wg.Wait()
+						return nil
+					default:
+					}
+
+					if err != nil {
+						return err
+					}
+					stream.In <- nodes.GetNode()
+				}
+			}
+			close(stream.In)
+			wg.Wait()
+
+			return nil
+		},
+	}
+
+	cmd.Flags().StringVar(&typ, "type", "", "the type of changeset to generate")
+	if err := cmd.MarkFlagRequired("type"); err != nil {
+		panic(err)
+	}
+	cmd.Flags().StringVar(&out, "out", "", "the directory to write changesets to")
+	if err := cmd.MarkFlagRequired("out"); err != nil {
+		panic(err)
+	}
+	cmd.Flags().IntVar(&limit, "limit", -1, "the version (inclusive) to halt generation at. -1 means no limit")
+	cmd.Flags().IntVar(&start, "start", 1, "the version (inclusive) to start generation at")
+
+	return cmd
+}
+
+func treeCommand() *cobra.Command {
+	var (
+		dbPath  string
+		genType string
+		limit   int64
+	)
+	cmd := &cobra.Command{
+		Use:   "tree",
+		Short: "build and save a Tree to disk, taking generated changesets as input",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			multiTree := iavl.NewMultiTree(dbPath, iavl.TreeOptions{StateStorage: true})
+			defer func(mt *iavl.MultiTree) {
+				err := mt.Close()
+				if err != nil {
+					log.Error().Err(err).Msg("failed to close db")
+				}
+			}(multiTree)
+
+			itr, err := getChangesetIterator(genType)
+			if err != nil {
+				return err
+			}
+
+			var i int64
+			var lastHash []byte
+			var lastVersion int64
+			start := time.Now()
+			for ; itr.Valid(); err = itr.Next() {
+				if err != nil {
```

Comment:
We need to catch the end of the iteration, but this code just returns on any generic error